src/bootstrap/download.rs

```diff
@@ -320,25 +320,44 @@ impl Config {
     }
 
     /// Returns whether the SHA256 checksum of `path` matches `expected`.
-    fn verify(&self, path: &Path, expected: &str) -> bool {
+    pub(crate) fn verify(&self, path: &Path, expected: &str) -> bool {
         use sha2::Digest;
 
         self.verbose(&format!("verifying {}", path.display()));
+
+        if self.dry_run() {
+            return false;
+        }
+
         let mut hasher = sha2::Sha256::new();
-        // FIXME: this is ok for rustfmt (4.1 MB large at time of writing), but it seems memory-intensive for rustc and larger components.
-        // Consider using streaming IO instead?
-        let contents = if self.dry_run() { vec![] } else { t!(fs::read(path)) };
-        hasher.update(&contents);
-        let found = hex::encode(hasher.finalize().as_slice());
-        let verified = found == expected;
-        if !verified && !self.dry_run() {
+
+        let file = t!(File::open(path));
+        const BUFFER_SIZE: usize = 4096; // read in chunks of 4KB
+        let mut reader = BufReader::with_capacity(BUFFER_SIZE, file);
```

Comment:
Why are we customizing the buffer size here vs. using the default in BufReader? If we *are* going to hard-code a buffer size, it seems like we might as well skip BufReader and just use `read(&mut buffer)` with a buffer we allocate here, which would also mean we can skip consume() since we wouldn't need any tracking of how much of that slice we've read.

Reply:
It was exactly like that(https://github.com/rust-lang/rust/pull/116635#discussion_r1356902119), I forgot removing the hard-coded size after moving to `fill_buf` and `consume`.